src/ngraph/runtime/gpu/cuda_emitter.cpp

```diff
@@ -1486,56 +1461,71 @@ size_t
     }
 
-    GPUAllocator allocator = this->m_primitive_emitter->get_memory_allocator();
-    size_t idx_strides = allocator.reserve_argspace(strides.data(), strides.size() * sizeof(int));
-    size_t idx_stride_magic =
-        allocator.reserve_argspace(stride_magic.data(), stride_magic.size() * sizeof(int));
-    size_t idx_stride_shift =
-        allocator.reserve_argspace(stride_shift.data(), stride_shift.size() * sizeof(int));
-    size_t idx_reduced_strides =
-        allocator.reserve_argspace(reduced_strides.data(), reduced_strides.size() * sizeof(int));
-
     size_t nthreads = shape_size(tensor_shape);
     constexpr const int nthreads_per_block = 32;
     int nblocks = 1 + ((static_cast<int>(nthreads) - 1) / nthreads_per_block);
 
-    // TODO: check if mutable is necessary
-    std::unique_ptr<gpu::primitive> ew_collective(new gpu::primitive{[=](void** inputs,
-                                                                         void** outputs) mutable {
-        void* strides_d = runtime::gpu::invoke_memory_primitive(m_ctx, idx_strides);
-        void* stride_magic_d = runtime::gpu::invoke_memory_primitive(m_ctx, idx_stride_magic);
-        void* stride_shift_d = runtime::gpu::invoke_memory_primitive(m_ctx, idx_stride_shift);
-        void* reduced_strides_d = runtime::gpu::invoke_memory_primitive(m_ctx, idx_reduced_strides);
+    auto args = this->m_primitive_emitter->add_kernel_args();
+    for (auto i = 0u; i < dtypes.size() - 1; i++)
+    {
+        args.add_placeholder(dtypes[i], "in" + std::to_string(i));
+    }
+    args.add_placeholder(dtypes.back(), "out0");
+    if (save_elementwise)
+    {
+        args.add_placeholder(dtypes.back(), "out1");
+    }
 
-        std::vector<void*> args_list;
-        for (auto i = 0u; i < dtypes.size() - 1; i++)
-        {
-            args_list.push_back(&inputs[i]);
-        }
-        args_list.push_back(&outputs[0]);
-        if (save_elementwise)
+    args.add("strides", strides)
+        .add("stride_magic", stride_magic)
+        .add("stride_shift", stride_shift)
+        .add("reduced_strides", reduced_strides)
+        .add("nthreads", nthreads);
+
+    auto compiled_kernel = m_ctx->compiled_kernel_pool->get(kernel_name.str());
+    if (compiled_kernel == nullptr)
+    {
+        codegen::CodeWriter writer;
+        CudaKernelBuilder::add_pod_typedefs(writer);
+        writer << include_helpers();
+        if (kernel)
         {
-            args_list.push_back(&outputs[1]);
+            CudaKernelBuilder::get_device_helper(writer, op, kernel, dtypes);
         }
-        args_list.push_back(&strides_d);
-        args_list.push_back(&stride_magic_d);
-        args_list.push_back(&stride_shift_d);
-        args_list.push_back(&reduced_strides_d);
-        args_list.push_back(&nthreads);
+        runtime::gpu::CudaKernelBuilder::get_kernel_signature(
+            writer, kernel_name.str(), args.get_input_signature());
+        CudaKernelBuilder::get_ew_collective_op(
```

Comment:
I don't disagree. Let me think on this a bit.

Reply:
@fengleitian, what do you think about this pseudocode:
https://github.com/NervanaSystems/ngraph/compare/cs/gpu_impl_kernelargs...cs/temp

The GPUKernelArgs object could be used in the kernels directly. This could provide some neat benefits. 

Assume `input_strides = {4, 2, 1}` for this example.

1. `args("input_strides", 1)` could emit different things depending on user-defined settings (possibly chosen by an auto-tuning framework). It could emit `input_strides1` which would have been defined in the argument list. Or it could emit the actual value of input_strides[1], `2` and in that case, `input_strides` wouldn't be an argument of the kernel but would instead be part of the kernel hash.

2. The hash for caching a kernel could be determined programmatically based on which arguments were used in the kernel.

3. Compile time checking of kernel argument use. If `input_strides` was not added to the argument list, using it in a kernel would give an error at ngraph compile time.

Point (1) can yield a performance improvement as the compiler can optimize operators on constant expressions (like `%` and `/`), and it also reduces the register count of a kernel, which could be quite high for kernels which have a large number of constant arguments.

Maybe not a high priority, but could be useful. Just a brainstorm I had last evening, let me know your thoughts.

Reply:
This looks good.